sql/core/src/main/scala/org/apache/spark/sql/Column.scala

```diff
@@ -39,10 +39,10 @@ private[sql] object Column {
 }
 
 /**
- * A [[Column]] where an [[Encoder]] has been given for the expected return type.
+ * A [[Column]] where an [[Encoder]] has been given for the expected input and return type.
  * @since 1.6.0
  */
-class TypedColumn[T](expr: Expression)(implicit val encoder: Encoder[T]) extends Column(expr)
+class TypedColumn[-T, U](expr: Expression, val encoder: Encoder[U]) extends Column(expr)
```

Comment:
Do we need a variance annotation here? Not much else uses it in the codebase. Also, it's not clear what the types T and U are (was is the "input type" of a column?).